### PR TITLE
[re-release][PT-1.12.1][SM] p4de launch, eval CVE

### PR DESCRIPTION
**Pull request by ShiboXing**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.12/py3/cu116/Dockerfile.sagemaker.gpu.os_scan_allowlist.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{
```
Comment (saimidu) [inline pytorch/training/docker/1.12/py3/cu116/Dockerfile.sagemaker.gpu.os_scan_allowlist.json:1]: This file needs to be renamed to `Dockerfile.ec2.gpu.os_scan_allowlist.json`

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu [hunk 2]
```diff
@@ -515,6 +515,16 @@ RUN  wget -nv https://github.com/rapidsai/rmm/archive/v${RMM_VERSION}.tar.gz \
 # Install ZeRO-2D Binary
 RUN pip install --no-cache-dir -U ${ZERO_2D_URL}
 
+# Install flash attention
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:518]: Why is this needed? do we have test to cover this new package?

Comment (rahul003) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:518]: Yes this is part of an S team goal for the release. It's a dependency of SMP going forward. This test touches that, but for proper test we'll need p4d. Running below test  for p4d will run with flash attention

https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py#L205

Comment (tejaschumbalkar) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:518]: is the test configured with p4d? I see the instance_type as `p3.16xlarge`

Comment (tejaschumbalkar) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:518]: Please make changes to the test configuration in order to test this functionality.

Comment (sheilaliuxl) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:518]: Yes, I chatted with Shibo offline that it should be `p4d.24xlarge` instead, thanks.

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py [hunk 3]
```diff
@@ -200,7 +200,7 @@ def test_smmodelparallel_gpt2_multigpu_singlenode(ecr_image, instance_type, sage
     framework, framework_version = get_framework_and_version_from_tag(ecr_image)
     if framework == "pytorch" and Version(framework_version) in SpecifierSet("==1.9.*"):
         pytest.skip("Skipping the test for PT1.9")
-    instance_type = "ml.p3.16xlarge"
+    instance_type = "ml.p4de.24xlarge"
```
Comment (sheilaliuxl) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py:203]: Can we change to `ml.p4d.24xlarge` instead, as `p4de` is new, and not widely available yet, based on offline discussion with our team, thanks.

Comment (ShiboXing) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py:203]: Will do

Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py]: Does the test needs to be ran for `p3.16xlarge` as well? Do we miss any coverage for existing functionality if we remove the `p3.16xlarge` instance type? If so, please parameterize `instance_type`  with both the instance types for pytest to pickup the configuration.

Comment (sheilaliuxl) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py]: Thanks for bringing it up, our team confirmed that `p4d` should be sufficient.

**Comment by ShiboXing** [comment]
SM standard: 040f7ab
SM EFA: 9df6d24
SM RC: 03109c6

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu [hunk 4]
```diff
@@ -515,6 +515,16 @@ RUN  wget -nv https://github.com/rapidsai/rmm/archive/v${RMM_VERSION}.tar.gz \
...
+WORKDIR /flash
+RUN git clone https://github.com/HazyResearch/flash-attention.git \
+    && cd flash-attention \
+    && git checkout 1d0b41be3b0fa759502e274779db6758617fc7b9 \
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:522]: nit: do we need to parameterize/reflect as an env var the commit id? We want to make sure this is reviewed once the dockerfile changes are carried forward to new versions.

**Review by tejaschumbalkar — COMMENTED** [review]
Please update the PR description

**Comment by tejaschumbalkar** [comment]
I think we should re-run the build and test, we saw failures on `test_pip_check` on #2504 a day ago due to sm-pysdk pinning the `packaging` version.

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu (hunk 4 above)
Comment (saimidu) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:522]: Changing this to an ARG defined at the top of the dockerfile, alongside the framework binaries, and adding a comment with reasoning.

Comment (sheilaliuxl) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:522]: I haven't got the reply from team yet, but the rubik team changes mirror that in https://code.amazon.com/packages/Rubik/blobs/tp_v2/--/containers/dockerfiles/Dockerfile.modelparallel.pytorch1.12.

Also the `ARG SMD_MODEL_PARALLEL_URL` points to the latest green rubik including those changes, so should have been validated there.

Comment (saimidu) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:522]: Thanks @sheilaliuxl! Will add a placeholder for now.

Comment (sheilaliuxl) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu]: A quick update: Flash attention did not have releases until recently, so we have to get the commit id for now.
In the next release, we can use their release tags if present.

**Comment by saimidu** [comment]
Test failures on sanity test jobs caused by https://github.com/aws/sagemaker-python-sdk/issues/3531

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu (hunk 4 above)
Comment (sheilaliuxl) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu]: Hi, I sent another PR: https://github.com/ShiboXing/deep-learning-containers/pull/4

This should fix the flash attention test. Note that most changes are from sync-to-head, and the only change is [this commit](https://github.com/ShiboXing/deep-learning-containers/commit/85dbac227580faa0138591f862513c2b35467fc1), thanks a lot for all your help!

**Comment by saimidu** [comment]
All tests pass:
SM standard: [f7627fa](https://github.com/aws/deep-learning-containers/pull/2498/commits/f7627fa4c4fc2d3db30a0746a3829e1b76b1d694)
SM EFA: [225ae80](https://github.com/aws/deep-learning-containers/pull/2498/commits/225ae80720521f5243733987f2608285fdfbcf6e)
SM RC: [b4fa87c](https://github.com/aws/deep-learning-containers/pull/2498/commits/b4fa87c9d6ee0b275909447a201f2ebfc39cdeca)

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py [hunk 5]
```diff
@@ -258,6 +258,83 @@ def test_smmodelparallel_gpt2_multigpu_singlenode(ecr_image, instance_type, sage
...
+
+@pytest.mark.usefixtures("feature_smmp_present")
+@pytest.mark.integration("smmodelparallel")
+@pytest.mark.model("gpt2")
+@pytest.mark.processor("gpu")
+@pytest.mark.skip_cpu
+@pytest.mark.skip_py2_containers
+@pytest.mark.skip_trcomp_containers
+@pytest.mark.parametrize("test_script, num_processes", [("train_gpt_simple.py", 8)])
+def test_smmodelparallel_gpt2_multigpu_singlenode_flashattn(ecr_image, instance_type, sagemaker_regions, test_script, num_processes):
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py:270]: Thank you @sheilaliuxl for this test.

**Review by saimidu — DISMISSED** [review]
Approved. Still need to resolve merge conflicts in pytorch/buildspec-1-12.yml, but all other contributions look good.

**Review by saimidu — APPROVED** [review]

**Review by sheilaliuxl — APPROVED** [review]
LGTM, thank you.

**Review by rahul003 — APPROVED** [review]
